### ISSUE TITLE
Default changed when opening recent projects

### ISSUE THREAD

**Issue by porsager** [zed-industries/zed#8651]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

When using cmd+alt+o to open recent projects they replace the currently open project instead of opening a new one.

Looking closer I noticed in the help text, that I now suddenly have to do cmd+enter to open in a new window, but I am unable to find any info about this change?

Can we please revert the original functionality and then make the cmd+enter option to have the new functionality instead of breaking the original behavior?

### Environment

Zed: v0.125.1 (Zed Preview)
OS: macOS 14.3.1
Memory: 32 GiB
Architecture: aarch64

### If applicable, add mockups / screenshots to help explain present your vision of the feature

<img width="606" alt="image" src="https://github.com/zed-industries/zed/assets/1288405/f573dfc4-3b3d-48b5-9403-d618d4f60a7f">

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.
If you only need the most recent lines, you can run the `zed: open log` command palette action to see the last 1000.

_No response_

**Comment by porsager** [comment]
This is actually way worse than the change in behaviour. Any unsaved changes will be discarded without warning when doing this! Just lost a lot of changes due to this. This is the number 1 thing any editor/app must not fail on. We're in 2024, we're not supposed to loose data unless we do something fucked up.

**Comment by SomeoneToIgnore** [comment]
This is my wrongdoing, sorry for breaking your workflow.
I have submitted the https://github.com/zed-industries/zed/pull/8673 to fix the prompt absence, thank you for the find.

After this data loss is fixed, we hopefully have only the broken workflow left: I've decided to make this change after two things:
* VSCode uses such behavior too, so my hopes were to have more people accepting this change as a reasonable default
It's also great to see that you've noticed an updated tooltip, so all options were still available.
* Zed still leaks memory on window open-close: https://github.com/zed-industries/zed/issues/7438 so it might be better to reduce such cases

I agree that it is not well to break user workflows, but we never agreed to have zero breakages on Preview, and this is the way I wanted to gather feedback — so far it's nice to see that not that many Preview users find it disrupting enough to create an issue or 👍 it.

I wonder, if we can adjust default keybindings for editors that lean to create new windows instead? I would imagine it's possible to have custom panel overrides and swap hotkeys for menu actions for them, would it help?

**Comment by porsager** [comment]
Sorry for the harsh wording re data loss - things like this are indeed something I must accept can happen when running preview, and I honestly forgot I was, so the tone in my reply was not warranted! Apologies!

Do you really think most users would prefer replacing current project? Perhaps it's a better solution to have 2 separate flow paths from the point of opening the prompt instead of at the point of selecting the project to open? It kind of relates to another issue I've created https://github.com/zed-industries/zed/issues/4357.

Basically have a "open project" dialog, and "switch to project" dialog?

**Comment by SomeoneToIgnore** [comment]
I have no clue tbh, but Nightly experiments on colleagues and now Preview experiments do not cause much backslash, so apparently it is true?
I am not standing strong on that hill though, and ready to revert it even if it goes on to Stable and only there people start complaining loudly.

Having a separate action sounds interesting, why not: we can have the `secondary` option in both to point at each other.
What if we keep things as is for ~2 versions or until a lot of negative feedback comes, and consider that second action?
To keep things less breaking this time, we can have the `switch to project` as a default `cmd-alt-o` option (keeping what's now) and allow other users to rebind it to the `open project` one?

**Comment by SomeoneToIgnore** [comment]
I've also added the way to change the defaults: https://github.com/zed-industries/zed/pull/8688
The default approach would be window reuse still, but now it's possible to make it work so that `enter`(`menu::Confirm`) opens projects in new windows.

I think now we're as flexible as we can be about opening the projects? 🙂 
All those are planned to be released as a new patch version in Zed Preview.

**Comment by SomeoneToIgnore** [comment]
Actually, one more good reason to revert back to default are collab projects: I, apparently, got used to the new way of working too fast and did not notice issues with that workflow, but more feedback from https://github.com/zed-industries/zed/pull/8798#issuecomment-1976031203 made me to consider it more, and now I would lean towards the old default after all.

To this Wednesday, we will have best of both worlds: override flexibility and non-broken defaults in Zed Stable.

**Comment by porsager** [comment]
It seems there's a regression on this in the latest zed version. Wanna take a look @SomeoneToIgnore ?

Zed: v0.161.2 (Zed)
OS: macOS 14.6.1
Memory: 32 GiB
Architecture: aarch64

**Comment by SomeoneToIgnore** [comment]
Oh, and we do have an override for the modal part of this:

https://github.com/zed-industries/zed/blob/1f3407e581aa4cc704c396eaf51c6cdf43bb1dea/assets/keymaps/default-macos.json#L399-L400

**Comment by porsager** [comment]
Hehe - override it is 🫠 Not the first time we cave to the MS overlords 😅

**Comment by porsager** [comment]
I'm sorry to bring this up again, but after a recent update it doesn't work again (i have the settings posted).

Zed: v0.166.1 (Zed)
OS: macOS 15.2.0
Memory: 32 GiB
Architecture: aarch64

**Comment by porsager** [comment]
I just realized that you also loose any terminal state when switching projects if not making new windows, so I'm even more confused this is the default now 🙈

**Comment by porsager** [comment]
ok - it seems to be all over the place.. The keybinding hack doesn't work anymore, and cmd+enter only works some of the time. 

I would really urge you to have the default as opening a new one - having all your active shells destroyed is very frustrating..

**Comment by SomeoneToIgnore** [comment]
Sorry for the inconveniences.

So far, it seems that VSCode lobby is strong, as no one else complains? 
That also seems the "right" thing to do given the VSCode keymap as a base.

We can fix forward the bugs that you're experiencing to make things better.

Can you elaborate on 

> keybinding hack doesn't work anymore

Is it the `{"create_new_window": true }` remap?


> cmd+enter only works some of the time

For me, it always works when I open recent projects, what do I miss?

**Comment by porsager** [comment]
I also haven't complained for a month even though it's been bugging me since my message in december :P It's one of those things where it happens during deep work maybe twice a week, and pausing while doing deep work to file a bug report doesn't happen. I know from several of my friends using zed it's happened to them too, so I don't think no complaints means it's not a frustration for other users.

I've spent some time now figuring out where the inconsistency lies.

When you don't have any file tabs open in a project it reuses the window regardless of your keybinding setting. This makes the behaviour hard to determine/inconsistent, and probably why I got the feeling the behaviour was "all over the place". The problem is that it doesn't care if you have shells open with running processes or logs you might need. If there are no file tabs it just discards those shells and opens the other project in the same window. Funny thing is that if you then (with the keybinding) use cmd+enter it does open a new window, so that caused me to use cmd+enter sometimes resulting in reusing the window when tabs were open. 😅

**Comment by SomeoneToIgnore** [comment]
I understand that there may happen a silent frustration somewhere, I was very against this myself for some time before got used to press `cmd` when needed.
The problem is, whatever we do with the defaults, the frustration won't go away but merely shift onto another group. 
So far, the most vocal group is covered it seems, and also the most logical for "VSCode keybinding base" thing is the default — unfortunately, defaults cannot satisfy everyone ergo the settings we try to have for the others.

So, let's fix the bugs with the overrides, when we understand what those are.

Here's me, trying to follow

> When you don't have any file tabs open in a project it reuses the window regardless of your keybinding setting.

and everything seem to work for me:

https://github.com/user-attachments/assets/e581951f-357d-4d34-a710-5ffd088aa473


As mentioned, I am used to the new model, so do not use the binding.
Happy to fix things when you can come up with a good repro.

**Comment by porsager** [comment]
Ok, I think I managed to make a proper screen recording of the issue now.

https://www.porsager.com/zed/zed_open_recent_issue.mov

**Comment by maxbrunsfeld** [comment]
I can't reproduce this bug on Zed Nightly. I always get a new window created when I do either:
* select a recent project with `cmd-enter`
* **or** open the recent projects picker in `create_new_window` mode, and then select a project with `enter`

For me, it makes no difference whether there are any tabs open. @SomeoneToIgnore I'm wondering if this was fixed recently?
@porsager what zed version are you on?

Also, I'm going to close this one out because we've decided to keep the behavior change that the original issue was objecting to. Maybe there is still a bug though? @porsager if you can still reproduce, could you open a separate bug report for that?

**Comment by SomeoneToIgnore** [comment]
I have not fixed it explicitly, but cannot repro this using the steps in the video too, unfortunately.
Not sure what's going on here...

**Comment by porsager** [comment]
I finally found out what the difference was.

I had the keyboard shortcut in the "Editor" context, so it was not about open or closed tabs at all, but just because I had the Shell focused in the case of no tabs, and since the shortcut was only bound to the editor context, it didn't have any effect in the shell causing the behaviour.

I have now changed context to `Workspace`, and it seems everything is good... phew!

So - final fix for others ending here:

Add this to keymap.json: **important that context is Workspace**
```json
{
  "context": "Workspace",
  "bindings": {
    "alt-cmd-o": ["projects::OpenRecent", { "create_new_window": true }],
  }
}
```

**Comment by SomeoneToIgnore** [comment]
Great, thanks for the reminder that we need to ask people to share their related settings/keymaps.

There's a

![Image](https://github.com/user-attachments/assets/705bdffb-f037-478e-9f36-5b3d10933750)

which might help with similar issues in the future:

![Image](https://github.com/user-attachments/assets/57b8cbfb-2a2b-4860-b4fa-bb6cfdcbf6ff)

**Comment by porsager** [comment]
Uh - very cool! Thanks

**Comment by PeterlitsZo** [comment]
I wanted to share some of my experiences and a bit of confusion I have around default behaviors. I use the Codex CLI for AI-powered coding across multiple projects. In VS Code, I don’t really use Open Recent — since it lacks search, I usually just open a new window with something like code ../foobar<TAB>. In Zed, I found a feature I really liked: just click the project name in the title bar, type part of another project’s name, and hit enter. But here’s what happened — my Codex session was completely gone. That felt pretty frustrating. I tried searching for 'project' and 'open recent' in the settings, but didn’t find anything. Yes, I know you can use Cmd-Opt-O to open a modal for opening projects, and there’s even support for configuring create_new_window — which is great. But what if I just want to click the project name in the title bar? I’m not sure if I’m the only one who finds this confusing. Another thing I’m not fully used to is not knowing whether my shell state will be discarded unless I actively check which windows I currently have open. And yeah, I get that VS Code’s Open Recent works the same way… but I don’t think VS Code’s default behavior is sane either.

cc @SomeoneToIgnore -- I want to open a issue to talk about it, but it is not a bug (maybe) and also not a feature request -- I just want to a talk about which behavior is better.
